src/Analysis/Ast/Impl/Analyzer/Caching/CacheFolders.cs

```diff
@@ -0,0 +1,62 @@
+﻿// Copyright(c) Microsoft Corporation
+// All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the License); you may not use
+// this file except in compliance with the License. You may obtain a copy of the
+// License at http://www.apache.org/licenses/LICENSE-2.0
+//
+// THIS CODE IS PROVIDED ON AN  *AS IS* BASIS, WITHOUT WARRANTIES OR CONDITIONS
+// OF ANY KIND, EITHER EXPRESS OR IMPLIED, INCLUDING WITHOUT LIMITATION ANY
+// IMPLIED WARRANTIES OR CONDITIONS OF TITLE, FITNESS FOR A PARTICULAR PURPOSE,
+// MERCHANTABILITY OR NON-INFRINGEMENT.
+//
+// See the Apache Version 2.0 License for specific language governing
+// permissions and limitations under the License.
+
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using System.Text;
+using Microsoft.Python.Core.IO;
+using Microsoft.Python.Core.OS;
+
+namespace Microsoft.Python.Analysis.Analyzer.Caching {
+    internal static class CacheFolders {
+        public static string GetCacheFilePath(string root, string moduleName, string content, IFileSystem fs) {
+            // Folder for all module versions and variants
+            // {root}/module_name/content_hash.pyi
+            var folder = Path.Combine(root, moduleName);
+
+            var filePath = Path.Combine(root, folder, $"{FileNameFromContent(content)}.pyi");
+            if (fs.StringComparison == StringComparison.Ordinal) {
+                filePath = filePath.ToLowerInvariant();
+            }
+            return filePath;
+        }
+
+        public static string GetCacheFolder(IOSPlatform platform) {
+            if (platform.IsWindows) {
+                return Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData), "Microsoft", "Python Language Server");
+            }
+
+            if (platform.IsMac) {
+                var macCachesFolder = Path.GetFullPath("~/Library/Caches");
+                return Path.Combine(macCachesFolder, "Microsoft/Python.Language.Server");
+            }
+
+            var linuxCachesFolder = Path.GetFullPath("~/.cache");
```

Comment:
Best to look up `$XDG_HOME_CACHE` first and use that before defaulting to `$HOME/.cache`.

Reply:
Err, `$XDG_CACHE_HOME`, sorry.